src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -102,6 +89,128 @@ pub struct ConfigSpace {
+            self.iovec.drop_descriptor_chain(&parsed_dc);
+            return Err(AddRxBufferError::BufferTooSmall);
+        }
+        self.parsed_descriptors.push_back(parsed_dc);
+        Ok(())
+    }
+
+    /// Returns the number of available `iovec` objects.
+    #[inline(always)]
+    fn len(&self) -> usize {
+        self.iovec.len()
+    }
+
+    /// Mark the first `size` bytes of available memory as used.
+    ///
+    /// The `size` should be the number of bytes previously written in the [`RxBuffers`]. This
+    /// means that `size` cannot be more than the currently available bytes in the underlying
+    /// `IoVecBufferMut`.
+    fn mark_used(&mut self, size: u32, metrics: &NetDeviceMetrics) {
+        // Since we were able to write a frame in guest memory, we should have at least one
+        // descriptor chain here. If not, we have a bug, so fail fast, since the device is
+        // fundamentally broken.
+        let mut parsed_dc = self.parsed_descriptors.pop_front().expect(
+            "net: internal bug. Mismatch between written frame size and available descriptors",
+        );
+
+        // We are trying to mark as used more bytes than a single descriptor. However without
+        // VIRTIO_NET_F_MRG_RXBUF negotiated (we don't support it currently) the guest can't handle
+        // frames spanning multiple DescriptorChains. There is nothing we can do now, just pass the
+        // descriptor back to the guest without claiming any written bytes
+        if size > parsed_dc.length {
+            parsed_dc.length = 0;
+            metrics.rx_fails.inc();
+        } else {
+            parsed_dc.length = size;
+            metrics.rx_bytes_count.add(size as u64);
+            metrics.rx_packets_count.inc();
+        }
+        self.header_set_num_buffers(1);
+        self.iovec.drop_descriptor_chain(&parsed_dc);
+        self.deferred_descriptor = Some(parsed_dc);
+    }
+
+    /// Write the number of descriptors used in VirtIO header
+    fn header_set_num_buffers(&mut self, nr_descs: u16) {
+        // We can unwrap here, because we have checked before that the `IoVecBufferMut` holds at
+        // least one buffer with the proper size, depending on the feature negotiation. In any
+        // case, the buffer holds memory of at least `std::mem::size_of::<virtio_net_hdr_v1>()`
+        // bytes.
+        self.iovec
+            .write_all_volatile_at(
+                &nr_descs.to_le_bytes(),
+                std::mem::offset_of!(virtio_net_hdr_v1, num_buffers),
+            )
+            .unwrap()
+    }
+
+    /// This will let the guest know that about all the `DescriptorChain` object that has been
+    /// used to receive a frame from the TAP.
+    fn finish_frame(&mut self, rx_queue: &mut Queue) -> bool {
+        match self.deferred_descriptor.take() {
+            Some(used_dc) => {
+                if let Err(err) = rx_queue.add_used(used_dc.head_index, used_dc.length) {
```

Comment:
Mh, I don't think this is completely right. The VirtIO spec, section 2.6.8 says

>  `len` [indicates] the total of bytes written into the buffer

so this shouldn't be `used_dc.length`, but `self.rx_bytes_read`, right?

Reply:
Ah, I see now, in `mark_used` we adjust `used_dc.length` to match the number of bytes we wrote. Nice! I do like that